draft/0002-subdocapi.md

```diff
@@ -0,0 +1,351 @@
+Adds a 'unique' element into an array (at which end is not defined or specified).
+Functions exactly like `extend_in`. There are some server-side constraints into
+which values can be used for unique operations - currently it is restricted to
+JSON primitives (so no compound items).
+
+* Throws `PathExists` if the unique element already exists
+* Throws `CantInsertValue` if the value is not a JSON primitive.
+* Throws `PathMismatch` if the array already contains non-primitive elements
+
+> Because there is too much variation in semantics between `add_unique_in` and
+> `extend_in`, they are separate methods rather than a single unified one.
+
+### `counter_in(docid, path, delta, create_parents, cas, ttl, persist_to, replicate_to)`
+Performs an atomic counter operation on a given path.
+
+* `delta`: The numeric delta, this is simply added (`+`) to the existing value.
+  This value may be signed to 'decrement' or unsigned to 'increment' the existing
+  value.
+
+See "Numeric Errors" below for a list of errors thrown by this operation.
+This method also returns the current value as a `MetaDocFragment`
+
+### `delete_in(docid, path, cas, ttl, persist_to, replicate_to)`
+Removes a path from the document
+
+## Multi-Path methods
+
+These methods take a document ID and a list of _specifications_ to perform. There are
+only two methods (`mutate_in` and `lookup_in`).
+
+The specification of each method is highly language specific, to use pseudo code, we will
+pretend there are two static helper classes which generate appropriate specs:
+
+### `LookupSpec`
+Creates lookup specifications. There are two 'static methods'
+
+* `LookupSpec.get(docid)` (takes the same arguments as `get_in`
+* `LookupSpec.exists(docid)` (takes the same arguments as `exists_in`)
+
+Both these objects should return a spec suitable for consumption by the `lookup_in`
+method.
+
+### `MutationSpec`
+Creates mutation specifications. There are several 'static methods' which may be used.
+These all function exactly like their single-path counterparts except that the document
+ID and cas are not provided (these are passed directly into `mutate_in`).
+
+* `MutationSpec.replace(path, value)`: See `replace_in`
+* `MutationSpec.upsert(path, value, create_parents)`. See `upsert_in`
+* `MutationSpec.insert(path, value, create_parents)`. See `insert_in`
+* `MutationSpec.extend(path, value_or_values, direction, create_parents)`. See `extend_in`
+* `MutationSpec.arrayinsert(path, value_or_values)`. See `arrayinsert_in`
+* `MutationSpec.addunique(path, value)`. See `addunique_in`
+* `MutationSpec.counter(path, delta)`. See `counter_in`
+* `MutationSpec.delete(path)`. See `delete_in`
+
+### `lookup_in(docid, lookup_specs...)`
+
+Retrieve (or check the existence of) multiple paths.
+
+Returns a list of MetaDocFragment objects. Individual objects may also
+contain non-success status codes (i.e. some paths may exist while others may fail).
```

Comment:
This is for the SDK to figure out. In Python for example, the return value is simply a list of (`status, value`) tuples; where a non-zero status indicates something other than success.

Note that the status is not attached to the fragment itself, but to the payload; meaning the `content` (or whatever is being used to contain the "value") is actually a collection of smaller results.